lightning/src/ln/channelmanager.rs

```diff
@@ -2589,6 +2594,64 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 		}
 	}
 
+	// Check that an inbound payment's `payment_data` field is sane. See `create_inbound_payment` and
+	// `create_inbound_payment_for_hash` for details on our checks.
+	fn verify_inbound_payment_data(&self, payment_hash: PaymentHash, payment_data: msgs::FinalOnionHopData) -> Result<Option<PaymentPreimage>, ()> {
+		let (iv_bytes, encrypted_metadata_bytes) = payment_data.payment_secret.0.split_at(16);
+		let mut chacha = ChaCha20::new(&self.inbound_payment_key[..], &iv_bytes[..12]);
+		let mut chacha_bytes = [0; 16];
+		chacha.process_in_place(&mut chacha_bytes);
+
+		let mut metadata_bytes: [u8; 16] = [0; 16];
+		for i in 0..16 {
+			metadata_bytes[i] = chacha_bytes[i] ^ encrypted_metadata_bytes[i];
+		}
+
+		let expiry = u64::from_be_bytes(metadata_bytes[8..].try_into().unwrap());
+		if expiry < self.highest_seen_timestamp.load(Ordering::Acquire) as u64 {
+			log_trace!(self.logger, "Failing HTLC with payment_hash {}: expired payment", log_bytes!(payment_hash.0));
```

Comment:
I'd feel 100x better if you only ever check the expiry/amount limits *after* checking relevant HMACs.